lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -842,10 +1059,108 @@ impl FileCacheObject {
+/// is the FileCacheObject itself, with it's own fso field representing
+/// the directory, and its contents being represented by FileSystemObject's
+/// in its contents field.
+///
+/// We account for an idiosyncrasy of the RDP protocol here: when fielding an
+/// IRP_MJ_DIRECTORY_CONTROL, RDP first expects to receive an entry for the "."
+/// directory, and next an entry for the ".." directory. Only after those two
+/// directories have been sent do we begin sending the actual contents of this
+/// directory (the contents field). (This is why we maintain dot_sent and dotdot_sent
+/// fields on each FileCacheObject)
+///
+/// Note that this implementation only makes sense in the case that this FileCacheObject
+/// is itself a directory (fso.file_type == FileType::Directory). We leave it up to the
+/// caller to ensure iteration makes sense in the given context that it's used.
+impl Iterator for FileCacheObject {
+    type Item = FileSystemObject;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // On the first call to next, return the "." directory
+        if !self.dot_sent {
+            self.dot_sent = true;
+            Some(FileSystemObject {
+                last_modified: self.fso.last_modified,
+                size: self.fso.size,
+                file_type: self.fso.file_type,
+                path: ".".to_string(),
+            })
+        } else if !self.dotdot_sent {
+            // On the second call to next, return the ".." directory
+            self.dotdot_sent = true;
+            Some(FileSystemObject {
+                last_modified: self.fso.last_modified,
+                size: 0,
+                file_type: FileType::Directory,
+                path: "..".to_string(),
+            })
+        } else {
+            // "." and ".." have been sent, now start iterating through
+            // the actual contents of the directory
+            if self.contents_i < self.contents.len() {
+                let i = self.contents_i;
+                self.contents_i += 1;
+                return Some(self.contents[i].clone());
+            }
+            None
+        }
+    }
+}
+
+struct FileCache {
+    cache: HashMap<u32, FileCacheObject>,
+}
+
+impl FileCache {
+    fn new() -> Self {
+        Self {
+            cache: HashMap::new(),
         }
     }
+
+    /// Insert a FileCacheObject into the file cache.
+    ///
+    /// If the file cache did not have this key present, [`None`] is returned.
+    ///
+    /// If the file cache did have this key present, the value is updated, and the old
+    /// value is returned. The key is not updated, though; this matters for
+    /// types that can be `==` without being identical.
+    fn insert(&mut self, file_id: u32, file: FileCacheObject) -> Option<FileCacheObject> {
+        self.cache.insert(file_id, file)
+    }
+
+    // TODO(isaiah): update get, get_mut, and remove to return RdpResult, because presumably none of them
```

Comment:
Is this handled in a future PR?

Reply:
No not yet. Deleted this comment and added https://github.com/gravitational/teleport/issues/13949